Make `BlockPopover` component public

gigitux opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

> [!NOTE]  
> This PR is just a draft to get feedback. If making sense to make the `BlockPopover` component public, I would suggest marking it as an unstable API and improving the documentation.

This PR makes the `BlockPopover` component public. 

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

Currently, a consumer can add the Gutenberg toolbar block with the `BlockToolbar` component. Hoverwer, the `BlockToolbar` isn't flexible, and it is complex to hide/show different buttons and text. 

Making `BlockPopover` public allows a consumer to create a custom `BlockToolbar` and customize it according to application needs.

This PR could be part of https://github.com/WordPress/gutenberg/issues/53874.

You can check https://github.com/woocommerce/woocommerce/issues/47323 and https://github.com/woocommerce/woocommerce/pull/47322 for a real-use case.



## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

gigitux committed:
make BlockPopover component public

youknowriad commented:
Hi there!, Thanks for opening this PR.

I would love more opinions on the PR (maybe @ellatrix) but I think making this component a public one is fine. But I'd probably make the two unstable props "private" before opening the component. (have a private version of the component and a public one)

youknowriad commented on packages/block-editor/README.md:
```diff
@@ -207,6 +207,10 @@ _Related_
 
+### BlockPopover
+
+Undocumented declaration.
```
Adding a JSDoc to the component helps with this. Also the component already has a README, so you might be able to reuse the README content.

gigitux committed:
create a public and a private version of blockpopover

gigitux committed:
update import

gigitux committed:
fix unit test

gigitux committed:
pass ref in the right way

gigitux commented:
I updated the PR, and I'm opening it to be reviewed. I:
- updated the documentation.
- created a public and a private version of the BlockPopover component.
- updated all the imports to ensure that the private version is used.

Let me know if I missed something! Thanks for the help!

youknowriad commented:
That looks good to me. I'd appreciate a second opinion @WordPress/gutenberg-core

noisysocks commented:
It's a useful component that I reach for a lot e.g. when building UIs for block supports. I imagine making it public might make it possible to build more advanced design tooling in plugins which is great. So I'm a 👍.

Before making it public let's be sure to check that there are no APIs (props, etc.) that we want to clean up first.

gigitux commented:
### Dev Note: BlockPopover Component Public

The `BlockPopover` component is now publicly accessible. For more details, you can check [the README](https://github.com/WordPress/gutenberg/blob/HEAD/packages/block-editor/src/components/block-popover/README.md).
